### DISCUSSION TITLE
Workaround for long proteins (>1024 residues) with ESM_1b

### DISCUSSION THREAD

**Discussion by salvatoreloguercio** [brianhie/evolocity#4] [Q&A] [answered]
Hi Brian,

First off, thanks for the great work with evolocity!

I have a quick follow-up question on the issue with longer proteins (>1024aa, https://github.com/brianhie/evolocity/issues/2) I closed a while ago - which I think is more appropriate for a Q&A section:

> The current workaround is to just divide the protein into 1022 residue windows (e.g., 1022 + before/after sequence tokens), run these through the model separately, then concatenate the output, but this is definitely a heuristic.

My understanding is that the concatenate operation would create embeddings of different size, depending of the number of chunks. Assuming that you will need embeddings of the same size for downstream tasks like classification, etc. - how the concatenated embeddings are reshaped to the length of the "base" embeddings, i.e. those for proteins <1022aa?

**Comment by brianhie** [comment, marked as answer]
Hi @salvatoreloguercio, ESM-1b will embed a sequence into an N x M matrix, where N is the sequence length and M is the embedding dimension (for ESM-1b, it's 1280). When N > 1022, we split the sequence into batches and then concatenate the resulting matrices along the sequence dimension.

For variable-length sequences, we take the mean embedding across the sequence dimension, resulting in a vector of length M for each sequence (see here: https://github.com/brianhie/evolocity/blob/070caec5427ab88aa7de4b9f48b919103b9227e2/evolocity/preprocessing/featurize_seqs.py#L118). I believe most approaches to neural biological sequence embedding take the same approach.

Hope that helps!
